aten/src/ATen/core/type.cpp

```diff
@@ -1079,6 +1079,299 @@ torch::jit::Function* ClassType::findForwardHook(const std::string& name) const
   return nullptr;
 }
 
+std::string ClassType::getPreHookErrorMessage(const std::string& pre_hook_name) const {
+  std::stringstream input_types;
+  const std::vector<Argument>& forward_args = getMethod("forward").getSchema().arguments();
+  for (int i = 1; i < forward_args.size(); ++i) {
+    input_types << forward_args[i].type()->annotation_str();
+    if (forward_args.size() - 1 != i) {
+      input_types << " ";
+    }
+  }
+  if (forward_args.size() == 1) {
+    input_types << "None";
+  }
+  std::string single_output = "";
+  if (forward_args.size() == 2 &&
+      forward_args[1].type()->cast<TupleType>() == nullptr) {
+    // if the output type is a single tuple, it needs to be wrapped in an outer tuple
+    // to match eager's behavior 
+    single_output = ", '" + forward_args[1].type()->annotation_str() + "',";
+  }
+  std::string pre_hook_schema =
+      pre_hook_name + "(self, input: Tuple[" + input_types.str() + "])";
+  std::string return_string =
+      "This error occured while scripting the forward pre-hook '" +
+      pre_hook_name + "' on module '" + name()->name() +
+      "'. If you did not want to script this pre-hook remove it from the "
+      "original NN module before scripting. Pre-hooks for module '" + 
+      name()->name() + "' are expected to have the following signature: " 
+      + pre_hook_schema + " with a return type of either 'None'" + 
+      single_output + " or 'Tuple[" + input_types.str() + "]'.";
+  return return_string;
+}
+
+std::string ClassType::getHookErrorMessage(const std::string& hook_name) const {
+  const FunctionSchema& forwardSchema = getMethod("forward").getSchema();
+  std::stringstream input_types;
+  // create expected input type string
+  const std::vector<Argument>& forward_args = forwardSchema.arguments();
+  for (int i = 1; i < forward_args.size(); ++i) {
+    input_types << forward_args[i].type()->annotation_str();
+    if (forward_args.size() - 1 != i) {
+      input_types << " ";
+    }
+  }
+  if (forward_args.size() == 1) {
+    input_types << "None";
+  }
+
+  // create expected output types string
+  const Argument& pre_output =
+      (forward_hooks_.size() == 1)
+          ? forwardSchema.returns()[0]
+          : forward_hooks_[forward_hooks_.size() - 2]->getSchema().returns()[0];
```

Comment:
You're right. I changed it to take the hook's index. It needs to be more specific than just the hook with the same name because the same hook can be called multiple time.